From a15dec4345ec3dbfa31b5d2708d7a0a5270b73f1 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Wed, 22 Jul 2026 23:52:59 -0400
Subject: [PATCH 218/268] atidxx: Fix some issues

Implements AmdD3D11CreateDeviceAndSwapChainExt to avoid crashes with native AGS. Primitive topology fixes
---
 dlls/atidxx64/atidxx.h      |   4 +-
 dlls/atidxx64/atidxx64.spec |   2 +-
 dlls/atidxx64/main.c        | 140 ++++++++++++++++++++++++++++++------
 3 files changed, 120 insertions(+), 26 deletions(-)

diff --git a/dlls/atidxx64/atidxx.h b/dlls/atidxx64/atidxx.h
index 9a152206919..3785a3c7e34 100644
--- a/dlls/atidxx64/atidxx.h
+++ b/dlls/atidxx64/atidxx.h
@@ -66,12 +66,12 @@ DECLARE_INTERFACE_(IAmdDxExt, IAmdDxExtInterface)
     THISCALLMETHOD_(IAmdDxExtInterface*,GetExtInterface)(THIS_ unsigned int iface) PURE;
 
 
-    THISCALLMETHOD_(HRESULT, IaSetPrimitiveTopology)(THIS_ D3D_PRIMITIVE_TOPOLOGY topology) PURE;
+    THISCALLMETHOD_(HRESULT, IaSetPrimitiveTopology)(THIS_ AmdDxExtPrimitiveTopology topology) PURE;
     THISCALLMETHOD_(HRESULT, IaGetPrimitiveTopology)(THIS_ AmdDxExtPrimitiveTopology *topology) PURE;
     THISCALLMETHOD_(HRESULT, SetSingleSampleRead)(THIS_ ID3D10Resource *res, BOOL single_sample) PURE;
     THISCALLMETHOD_(HRESULT, SetSingleSampleRead11)(THIS_ ID3D11Resource *res, BOOL single_sample) PURE;
     THISCALLMETHOD_(HRESULT, QueryFeatureSupport)(THIS_ unsigned int feature_token, void *data, unsigned int data_size) PURE;
-    THISCALLMETHOD_(HRESULT, IaSetPrimitiveTopologyCtx)(THIS_ unsigned int topology, ID3D11DeviceContext *ctx) PURE;
+    THISCALLMETHOD_(HRESULT, IaSetPrimitiveTopologyCtx)(THIS_ AmdDxExtPrimitiveTopology topology, ID3D11DeviceContext *ctx) PURE;
     THISCALLMETHOD_(HRESULT, IaGetPrimitiveTopologyCtx)(THIS_ AmdDxExtPrimitiveTopology *topology, ID3D11DeviceContext *ctx) PURE;
 };
 #undef INTERFACE
diff --git a/dlls/atidxx64/atidxx64.spec b/dlls/atidxx64/atidxx64.spec
index 4012dcb0303..1c59308f080 100644
--- a/dlls/atidxx64/atidxx64.spec
+++ b/dlls/atidxx64/atidxx64.spec
@@ -1,6 +1,6 @@
 @ stub AmdD3D11CreateDevice
 @ stub AmdD3D11CreateDeviceAndSwapChain
-@ stub AmdD3D11CreateDeviceAndSwapChainExt()
+@ stdcall AmdD3D11CreateDeviceAndSwapChainExt(ptr long ptr long ptr long long ptr ptr ptr ptr ptr ptr)
 @ stdcall AmdD3D11CreateDeviceExt(ptr long ptr long ptr long long ptr ptr ptr ptr)
 @ cdecl AmdDxExtCreate11(ptr ptr)
 @ stub AmdDxExtCreate
diff --git a/dlls/atidxx64/main.c b/dlls/atidxx64/main.c
index de38ad36a6c..88f8d9bee63 100644
--- a/dlls/atidxx64/main.c
+++ b/dlls/atidxx64/main.c
@@ -45,12 +45,6 @@
 WINE_DEFAULT_DEBUG_CHANNEL(atidxx);
 
 static HMODULE d3d11_module;
-static HRESULT (WINAPI *pD3D11CreateDevice)(IDXGIAdapter *adapter, D3D_DRIVER_TYPE driver_type,
-        HMODULE swrast, UINT flags, const D3D_FEATURE_LEVEL *feature_levels, UINT levels,
-        UINT sdk_version, ID3D11Device **device_out, D3D_FEATURE_LEVEL *obtained_feature_level,
-        ID3D11DeviceContext **immediate_context);
-
-typedef void (*vtable_ptr)(void);
 
 static HRESULT load_d3d11(void)
 {
@@ -72,6 +66,11 @@ HRESULT WINAPI AmdD3D11CreateDeviceExt(IDXGIAdapter *adapter, D3D_DRIVER_TYPE dr
         D3D_FEATURE_LEVEL *obtained_feature_level, ID3D11DeviceContext **immediate_context, void *unk)
 {
     HRESULT ret;
+    static HRESULT (WINAPI *pD3D11CreateDevice)(IDXGIAdapter *adapter, D3D_DRIVER_TYPE driver_type,
+            HMODULE swrast, UINT flags, const D3D_FEATURE_LEVEL *feature_levels, UINT levels,
+            UINT sdk_version, ID3D11Device **device_out, D3D_FEATURE_LEVEL *obtained_feature_level,
+            ID3D11DeviceContext **immediate_context);
+
     FIXME("%p semi-stub\n", unk);
 
     if ((ret = load_d3d11()))
@@ -84,6 +83,30 @@ HRESULT WINAPI AmdD3D11CreateDeviceExt(IDXGIAdapter *adapter, D3D_DRIVER_TYPE dr
             device_out, obtained_feature_level, immediate_context);
 }
 
+HRESULT WINAPI AmdD3D11CreateDeviceAndSwapChainExt(IDXGIAdapter *adapter, D3D_DRIVER_TYPE driver_type,
+        HMODULE swrast, UINT flags, const D3D_FEATURE_LEVEL *feature_levels, UINT levels,
+        UINT sdk_version, const DXGI_SWAP_CHAIN_DESC *swapchain_desc, IDXGISwapChain **swapchain,
+        ID3D11Device **device_out, D3D_FEATURE_LEVEL *obtained_feature_level, ID3D11DeviceContext **immediate_context,
+        void *unk)
+{
+    HRESULT ret;
+    static HRESULT (WINAPI *pD3D11CreateDeviceAndSwapchain)(IDXGIAdapter *adapter, D3D_DRIVER_TYPE driver_type,
+            HMODULE swrast, UINT flags, const D3D_FEATURE_LEVEL *feature_levels, UINT levels,
+            UINT sdk_version, const DXGI_SWAP_CHAIN_DESC *swapchain_desc, IDXGISwapChain **swapchain,
+            ID3D11Device **device_out, D3D_FEATURE_LEVEL *obtained_feature_level, ID3D11DeviceContext **immediate_context);
+
+    FIXME("%p semi-stub\n", unk);
+
+    if ((ret = load_d3d11()))
+        return ret;
+
+    if (!pD3D11CreateDeviceAndSwapchain)
+        pD3D11CreateDeviceAndSwapchain = (void *)GetProcAddress(d3d11_module, "D3D11CreateDeviceAndSwapchain");
+
+    return pD3D11CreateDeviceAndSwapchain(adapter, driver_type, swrast, flags, feature_levels, levels, sdk_version,
+                                          swapchain_desc, swapchain, device_out, obtained_feature_level, immediate_context);
+}
+
 /*
     Ext Ifaces to implement:
     field_0x170 = 0x11
@@ -182,9 +205,9 @@ unsigned int __thiscall AmdDxExt_Release(IAmdDxExt *iface)
 {
     AmdDxExt *This = impl_from_IAmdDxExt(iface);
     LONG ref = InterlockedDecrement(&This->ref);
-    if(ref == 0)
+    if (!ref)
     {
-        if(This->is_d3d11)
+        if (This->is_d3d11)
         {
             ID3D11Device_Release(This->device);
             ID3D11DeviceContext_Release(This->context);
@@ -219,7 +242,7 @@ IAmdDxExtInterface* __thiscall AmdDxExt_GetExtInterface(IAmdDxExt *ext, unsigned
     IAmdDxExtInterface *ret = NULL;
     TRACE("%p %x\n", ext, iface);
 
-    switch(iface)
+    switch (iface)
     {
         case 0x2:
             ret = (IAmdDxExtInterface *)&This->IAmdDxExtQuadBufferStereo_iface;
@@ -245,26 +268,82 @@ IAmdDxExtInterface* __thiscall AmdDxExt_GetExtInterface(IAmdDxExt *ext, unsigned
     return ret;
 }
 
+static D3D_PRIMITIVE_TOPOLOGY convert_topology_d3d(AmdDxExtPrimitiveTopology top)
+{
+    switch (top)
+    {
+        case AmdDxExtPrimitiveTopology_LineList:
+        case AmdDxExtPrimitiveTopology_LineListAdj:
+        case AmdDxExtPrimitiveTopology_TriangleStrip:
+        case AmdDxExtPrimitiveTopology_TriangleStripAdj:
+        case AmdDxExtPrimitiveTopology_TriangleList:
+        case AmdDxExtPrimitiveTopology_TriangleListAdj:
+        case AmdDxExtPrimitiveTopology_LineStrip:
+        case AmdDxExtPrimitiveTopology_LineStripAdj:
+        case AmdDxExtPrimitiveTopology_PointList:
+        case AmdDxExtPrimitiveTopology_Undefined:
+            return (D3D_PRIMITIVE_TOPOLOGY)top;
+        /* FIXME: what are the number of verticies in patch list?
+         * Vulkan doesn't support quads so we can't implement that */
+        default:
+            ERR("Failed to map topology %u\n", top);
+            return D3D_PRIMITIVE_TOPOLOGY_UNDEFINED;
+    }
+}
+
 DEFINE_THISCALL_WRAPPER(AmdDxExt_IaSetPrimitiveTopology, 8)
-HRESULT __thiscall AmdDxExt_IaSetPrimitiveTopology(IAmdDxExt *ext, D3D_PRIMITIVE_TOPOLOGY topology)
+HRESULT __thiscall AmdDxExt_IaSetPrimitiveTopology(IAmdDxExt *ext, AmdDxExtPrimitiveTopology topology)
 {
     AmdDxExt *This = impl_from_IAmdDxExt(ext);
-    TRACE("%p %u\n", ext, topology);
+    FIXME("%p %u semi-stub\n", ext, topology);
 
-    if(This->is_d3d11)
-    {
-        ID3D11DeviceContext_IASetPrimitiveTopology(This->context, topology);
-    }
+    if (topology >= AmdDxExtPrimitiveTopology_Max) return E_INVALIDARG;
+    /* these don't exist in vulkan */
+    if (topology == AmdDxExtPrimitiveTopology_ExtQuadList) return E_NOTIMPL;
+    if (topology == AmdDxExtPrimitiveTopology_ExtScreenRectList) return E_NOTIMPL;
+
+    if (This->is_d3d11)
+        ID3D11DeviceContext_IASetPrimitiveTopology(This->context, convert_topology_d3d(topology));
 
     return S_OK;
 }
 
+static AmdDxExtPrimitiveTopology convert_topology_ext(D3D_PRIMITIVE_TOPOLOGY top)
+{
+    switch (top)
+    {
+        case D3D_PRIMITIVE_TOPOLOGY_TRIANGLESTRIP:
+        case D3D_PRIMITIVE_TOPOLOGY_TRIANGLESTRIP_ADJ:
+        case D3D_PRIMITIVE_TOPOLOGY_LINELIST:
+        case D3D_PRIMITIVE_TOPOLOGY_LINELIST_ADJ:
+        case D3D_PRIMITIVE_TOPOLOGY_POINTLIST:
+        case D3D_PRIMITIVE_TOPOLOGY_LINESTRIP:
+        case D3D_PRIMITIVE_TOPOLOGY_LINESTRIP_ADJ:
+        case D3D_PRIMITIVE_TOPOLOGY_TRIANGLELIST:
+        case D3D_PRIMITIVE_TOPOLOGY_TRIANGLELIST_ADJ:
+        case D3D_PRIMITIVE_TOPOLOGY_UNDEFINED:
+            return (AmdDxExtPrimitiveTopology)top;
+        /* the remaining are patchlist, but not sure how they map to AmdDxExt */
+        default:
+            ERR("Failed to map topology %u\n", top);
+            return AmdDxExtPrimitiveTopology_Undefined;
+    }
+}
+
 DEFINE_THISCALL_WRAPPER(AmdDxExt_IaGetPrimitiveTopology, 8)
 HRESULT __thiscall AmdDxExt_IaGetPrimitiveTopology(IAmdDxExt *ext, AmdDxExtPrimitiveTopology *topology)
 {
-    FIXME("%p %p stub\n", ext, topology);
+    AmdDxExt *This = impl_from_IAmdDxExt(ext);
+    D3D11_PRIMITIVE_TOPOLOGY d3d_topology;
+    FIXME("%p %p semi-stub\n", ext, topology);
 
-    return E_NOTIMPL;
+    if (This->is_d3d11)
+    {
+        ID3D11DeviceContext_IAGetPrimitiveTopology(This->context, &d3d_topology);
+        *topology = convert_topology_ext(d3d_topology);
+    }
+
+    return S_OK;
 }
 
 DEFINE_THISCALL_WRAPPER(AmdDxExt_SetSingleSampleRead, 12)
@@ -292,19 +371,30 @@ HRESULT __thiscall AmdDxExt_QueryFeatureSupport(IAmdDxExt *iface, unsigned int f
 }
 
 DEFINE_THISCALL_WRAPPER(AmdDxExt_IaSetPrimitiveTopologyCtx, 12)
-HRESULT __thiscall AmdDxExt_IaSetPrimitiveTopologyCtx(IAmdDxExt *iface, unsigned int topology, ID3D11DeviceContext *ctx)
+HRESULT __thiscall AmdDxExt_IaSetPrimitiveTopologyCtx(IAmdDxExt *iface, AmdDxExtPrimitiveTopology topology, ID3D11DeviceContext *ctx)
 {
-    FIXME("%p %u %p stub\n", iface, topology, ctx);
+    FIXME("%p %u %p semi-stub\n", iface, topology, ctx);
 
-    return E_NOTIMPL;
+    if (topology >= AmdDxExtPrimitiveTopology_Max) return E_INVALIDARG;
+    /* these don't exist in vulkan */
+    if (topology == AmdDxExtPrimitiveTopology_ExtQuadList) return E_NOTIMPL;
+    if (topology == AmdDxExtPrimitiveTopology_ExtScreenRectList) return E_NOTIMPL;
+
+    ID3D11DeviceContext_IASetPrimitiveTopology(ctx, convert_topology_d3d(topology));
+
+    return S_OK;
 }
 
 DEFINE_THISCALL_WRAPPER(AmdDxExt_IaGetPrimitiveTopologyCtx, 12)
 HRESULT __thiscall AmdDxExt_IaGetPrimitiveTopologyCtx(IAmdDxExt *iface, AmdDxExtPrimitiveTopology *topology, ID3D11DeviceContext *ctx)
 {
-    FIXME("%p %p %p stub\n", iface, topology, ctx);
+    D3D_PRIMITIVE_TOPOLOGY d3d_topology;
+    FIXME("%p %p %p semi-stub\n", iface, topology, ctx);
 
-    return E_NOTIMPL;
+    ID3D11DeviceContext_IAGetPrimitiveTopology(ctx, &d3d_topology);
+    *topology = convert_topology_ext(d3d_topology);
+
+    return S_OK;
 }
 
 /* our release method handles destruction */
@@ -483,6 +573,9 @@ void __thiscall AmdDxExtMultidrawIndirect_GetVersion(IAmdDxExtMultidrawIndirect
 {
     AmdDxExt *This = impl_from_IAmdDxExtMultidrawIndirect(iface);
     FIXME("%p %p semi-stub!\n", iface, version);
+    /* uVar3 = (version.major << 10 | version.minor) << 12 */
+    /* 0x401fff < uVar3 -> multi draw indirect count is supported */
+    /* 0x403fff < uVar3 -> different version of multidraw indirect that has different offsets (maybe for crossfire) */
     version->majorVersion = 1;
     /* needed for indirect count */
     version->minorVersion = This->multi_draw_indirect_count ? 2 : 0;
@@ -518,6 +611,7 @@ HRESULT __thiscall AmdDxExtMultidrawIndirect_MultiDrawIndexedIndirect(IAmdDxExtM
     return S_OK;
 }
 
+/* Taken from Proton's amd_ags_x64 implementation */
 static unsigned int get_max_draw_count(ID3D11Buffer *buffer, unsigned int offset, unsigned int stride, unsigned int size)
 {
     D3D11_BUFFER_DESC desc;
@@ -525,7 +619,7 @@ static unsigned int get_max_draw_count(ID3D11Buffer *buffer, unsigned int offset
 
     ID3D11Buffer_GetDesc(buffer, &desc);
 
-    if(offset >= desc.ByteWidth)
+    if (offset >= desc.ByteWidth)
     {
         WARN("Offset %u, buffer size %u.", offset, desc.ByteWidth);
         return 0;
-- 
2.55.0

